### PR TITLE
Delay submission of Upload DIP to ArchivesSpace after pair reviewing

### PR DESCRIPTION
**Pull request by sevein**
This pull request is an alternative to #1177 that does not require workflow modifications by making changes in the user interface so the job "Upload DIP" is not executed until the user is presented with the pair reviewing page and submits the form.

Once submitted, the user is forwarded to the ingest grid page and it's prompted with the "Choose Config for ArchivesSpace DIP Upload". Further work could be done to remove this step but that would require a database migration which is not ideal in this work because it's targeting a point release too.

More details have been included in the description of the commits.

This is connected to #1112.

### CODE REVIEW

File: src/dashboard/src/contrib/mcp/client.py
Comment (ross-spencer) [inline src/dashboard/src/contrib/mcp/client.py]: We can define this as a custom exception, `MCPNoJobFoundException` maybe? This would mean we can be more explicit about the exception we want to handle in calling functions. This reduces the risk of hiding other exceptions/bugs that might be in calling functions.

File: src/dashboard/src/components/ingest/views_as.py
Comment (ross-spencer) [inline src/dashboard/src/components/ingest/views_as.py]: I think we want to try to avoid catching generic exceptions as much as possible. See below.

File: src/dashboard/src/media/js/ingest.js
Comment (ross-spencer) [inline src/dashboard/src/media/js/ingest.js]: I don't know how likely it is we need to handle a situation where the sip uuid isn't found? i.e. `this.model.sip.get('uuid')` but it seems we can avoid this repetition in calling `get(...)` twice, and increase the potential for handling that by assigning it to a variable first before assigning it in this object.

**Review by ross-spencer — CHANGES_REQUESTED** [review]
Some small changes requested here @sevein. It looks like a nice solution compared to having to introduce a migration for a point release.

File: src/dashboard/src/components/ingest/views_as.py
Comment (sevein) [reply src/dashboard/src/components/ingest/views_as.py]: I'm not sure. What I'm doing is a common pattern in model-view-controller. This is a view and the main goal is to output information that can be understood by the user. If I don't handle a particular exception then it would bubble up and be caught by the outer middleware which would present a generic error page - I'm trying to avoid that.

File: src/dashboard/src/contrib/mcp/client.py
Comment (sevein) [reply src/dashboard/src/contrib/mcp/client.py]: I didn't do it because I thought it didn't have an utility, do you think it's still needed after reading my response above? I can add it if you wish. The exception message would still be logged but at the moment we don't want to do anything specific about this error.

File: src/dashboard/src/media/js/ingest.js
Comment (sevein) [reply src/dashboard/src/media/js/ingest.js]: I've declared the variable (`unitId`) as suggested but I don't know how I could handle and report the error. On one hand I don't think a backbone model would make it this far without an ID. But also I don't think there's logic in place at the moment to report the error to the user. What would you do?

```javascript
if (!this.model.sip.get('uuid')) {
    // Would you return?
    // Log to the console?
    // ...?
}
```

**Comment by sevein** [comment]
Ready for review again!

**Comment by sevein** [comment]
Reminder: once merged, I need to port this to stable/1.7.x. It'll conflict in `ingest.js`.

File: src/dashboard/src/contrib/mcp/client.py
Comment (ross-spencer) [reply src/dashboard/src/contrib/mcp/client.py]: I can't see the original message in this thread but I remember the gist was that we want to return something simple that the user can understand. I had imagined that you could make the exception specific, because at the end of the day, it is, and then from the dashboard still return a generic message to the user, with a specific message to the logger. I wonder would that work? Or is it likely not robust enough for what a user might experience in the UI?

File: src/dashboard/src/media/js/ingest.js
Comment (ross-spencer) [inline src/dashboard/src/media/js/ingest.js:296]: Thanks @sevein. In response to your other message about how you'd handle a potential exception here, I don't think we need to do that at this point. You're right that it seems highly unlikely.

File: src/dashboard/src/media/js/ingest.js
Comment (ross-spencer) [inline src/dashboard/src/media/js/ingest.js:315]: What's the Artefactual best practice for leaving `TODO` in code? A suggestion I've made in the past is to open a new GitHub issue where this type of test might be out of scope? It could also be left as a note in the acceptance criteria?

**Review by ross-spencer — COMMENTED** [review]
Feel free to make these changes, or leave as is @sevein. A few discussion points we can pick up on another time. 

I'm away until Monday after today.  If you make any changes you need approval on and get them in later today I can make sure I loop before COP. Otherwise, can pick up Monday if needed. :+1:

File: src/dashboard/src/media/js/ingest.js
Comment (sevein) [reply src/dashboard/src/media/js/ingest.js:315]: I'll add a note, good idea, thanks.

File: src/dashboard/src/contrib/mcp/client.py
Comment (sevein) [reply src/dashboard/src/contrib/mcp/client.py]: If I understand correctly what you mean, yes it would work, you can definitely create as many custom exceptions as you want and raise them and still result in the same behaviour where I have a catch-all up in the stack as I'm doing.

However I don't know if I see the benefit, to me it's like a `if False: ...` ([we used to have some](https://github.com/artefactual/archivematica-history/blob/407bcb082f382dcfc54c5ec9b22a819ee6903402/src/MCPServer/lib/archivematicaMCP.py#L241)) which I'd remove it right away because it's dead code. Wouldn't this be similar? Excepting that it's harder to reason about because you also have to check its usage outside the function? It's like a fancy goto.

In any case, I can see the value of making this available for a potential use case in the future. Let me know if this would work for you:

```python

# In contrib.mcp.client

class NoJobFoundError(Exception):
    def __init__(self, *args, **kwargs):
        try:
            message = args[0]
        except IndexError:
            message = "No job was found"
        super(NoJobFoundError, self).__init__(message)


# Usage:
raise NoJobFoundError()
raise NoJobFoundError("custom error message")
```

Comment (ross-spencer) [reply src/dashboard/src/contrib/mcp/client.py]: This looks good @sevein :)

File: src/dashboard/src/components/ingest/views_as.py
Comment (sevein) [inline src/dashboard/src/components/ingest/views_as.py:242]: @ross-spencer thanks I've added `NoJobFoundError`.

To double check, should the snippet above remain the same?

Comment (ross-spencer) [reply src/dashboard/src/components/ingest/views_as.py:242]: :+1:

**Comment by sevein** [comment]
e209c8eafb0855964d542206079b36eac5e66cc3 and bc4c887db40bb3d9cb723eac4580b8bea72c25c3 in `stable/1.7.x`.

My mistake was to target qa/1.x first, I should have started with stable/1.7.x.

**Pull request merged**
